tests/pd/mock/server.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::net::ToSocketAddrs;
+
+use futures;
+use futures::Future;
+
+use grpc::error::GrpcError;
+use grpc::futures_grpc::{GrpcFutureSend, GrpcStreamSend};
+
+use kvproto::pdpb::*;
+use kvproto::pdpb_grpc::{PDAsync, PDAsyncServer};
+
+use super::Case;
+use super::case::Result;
+
+pub struct Server {
+    _server: PDAsyncServer,
+}
+
+impl Server {
+    pub fn run<A, H, C>(addr: A, handler: Arc<H>, case: Option<Arc<C>>) -> Server
+        where A: ToSocketAddrs,
+              H: Case + Send + Sync + 'static,
+              C: Case + Send + Sync + 'static
+    {
+        let m = Mock {
+            handler: handler,
+            case: case,
+        };
+        Server { _server: PDAsyncServer::new(addr, Default::default(), m) }
+    }
+}
+
+fn try_takeover<F, R, A: Case, B: Case>(mock: &Mock<A, B>, f: F) -> GrpcFutureSend<R>
+    where R: Send + 'static,
+          F: Fn(&Case) -> Option<Result<R>>
+{
+    if let Some(ref case) = mock.case {
+        match f(case.as_ref()) {
+            Some(Ok(resp)) => return futures::future::ok(resp).boxed(),
+            Some(Err(err)) => return futures::future::err(err).boxed(),
+            _ => (),
+        }
+    }
+
+    match f(mock.handler.as_ref()) {
+        Some(Ok(resp)) => futures::future::ok(resp).boxed(),
+        Some(Err(err)) => futures::future::err(err).boxed(),
+        _ => futures::future::err(GrpcError::Other("unimpl")).boxed(),
+    }
+}
+
+#[derive(Debug)]
+struct Mock<C: Case, H: Case> {
+    handler: Arc<H>,
```

Comment:
The handler here is not only for `Service` but also for other `Cases`. `handler` is designed for providing default rpc stub, and `case` is for overriding, so that a more complex case can be built-up those simple cases.